plutus-tx-plugin/test/TH/traceRepeatedly.plc.golden

```diff
@@ -1 +1,5 @@
-[Making my first int, Making my second int, Adding them up]
\ No newline at end of file
+[ Making my first int
+, Making my second int
+, Making my first int
```

Comment:
@michaelpj @effectfully any idea why this is shown twice? Looks like a bug to me.

Reply:
Seems like the inliner considers `i1` and `i2` safe to inline, @michaelpj is in a better position to answer.

Reply:
> Because the let-binding is non-strict and it's used twice.

So there's no memoization in PLC evaluation? That's interesting

Reply:
Non-strict bindings are compiled like this:
```
let nonstrict x = y in ... y ...
==>
let strict x = \_ -> y in ... (y ()) ...
```
it's ugly, and I somewhat wish we'd made strict bindings the default. It wouldn't have been that surprising and this behaviour is itself pretty surprising.

Reply:
I've managed to forget that we do it this way. I thought let-expressions were strict by default. That certainly makes it harder to explain to people the semantics of the language.

Reply:
Yes, and results in more code bloat, which is one of the biggest problems so far. I'm more worried about semantics though. "By default bindings are lazy in a strict language compiled from a fully lazy language" -- good luck trying to figure out what the heck this all means.

Although if we pretended that Haskell had no influence on Plutus apart from the syntax, it would be equally unclear how to move a piece of code from off-chain to on-chain or vice versa. Maybe it is already unclear, though.

Those strictness concerns do suck.

Reply:
So couldn't we say that our `let`s are call-by-name and not Haskell's call-by-need? The problem is only observable when using `trace` right?

Reply:
We do make some compromises with Haskell's semantics. This is one of them. Maybe it was a bad idea. I thought that Haskellers would be too used to being able to move around let-bindings freely without consequences, which would no longer be true if they were strict by default...